File: Java/arrays/TrappingRainwater.java
// https://leetcode.com/problems/trapping-rain-water/description/

// Problem 42: Trapping Rainwater

public class TrappingRainwater { // in leetcode instead of this there is class solution 
    
        public static int trap(int[] height) {
            int n=height.length;int c1=1;int c2=1;int WT=0;int width=1;
            int LM[]=new int[n];
            LM[0]=height[0]; // LeftMax starts from zero

            //LeftMax boundary calculation using helper array

            for(int i=1;i<n;i++){
                LM[i]=Math.max(height[i],LM[i-1]);
            }

            // RightMax boundary calculation using helper array

            int RM[]=new int[n];
            RM[n-1]=height[n-1];
            for(int i=n-2;i>=0;i--){
                RM[i]=Math.max(height[i],RM[i+1]);
            }

            

            for(int i=0;i<n;i++){
                //WaterTrapped=(waterLevel-height)* width for one block =(min(LM,RM) - height) *width for one block
                    WT+=( (Math.min(LM[i],RM[i]))-height[i]) * width;
            }
            return WT;
        }
        public static void main(String[] args) {
            int height[]={4,2,0,6,3,2,5};
             int ans=trap(height); 
             System.out.println(ans);
        }
}

File: C++/arrays/TrappingRainWater.cpp
/*
Problem: Trapping Rain Water
Topic: Arrays, Two-Pointer
Description:
You are given an array `height` representing the elevation map where the width of each bar is 1.
Calculate how much water it is able to trap after raining.

Example:
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: 6 units of water can be trapped between the bars.

Constraints:
- 1 <= height.length <= 10^5
- 0 <= height[i] <= 10^5

Time Complexity: O(n)
Space Complexity: O(1)
*/

/*
Algorithm intuition (short):
- Water trapped at each bar depends on the maximum height to its left and right.
- Use two pointers: left and right, with leftMax and rightMax to track the highest walls.
- Move the pointer with smaller height:
    - If current height < leftMax/rightMax, add trapped water: leftMax/rightMax - height
    - Otherwise, update leftMax/rightMax
- Continue until pointers meet. Accumulate water along the way.
*/

#include <bits/stdc++.h>
using namespace std;

class TrappingRainWater {
public:
    int trap(vector<int>& height) {
        if (height.empty()) return 0;// if no bars the answer is zero

        int left = 0;
        int right = height.size() - 1;
        int leftMax = height[left];
        int rightMax = height[right];
        int water = 0;

        // Always move the pointer with the smaller max height
        // Water trapped at a bar = min(leftMax, rightMax) - currentHeight
        // By moving the smaller side, we ensure correct calculation and O(n) time

        while (left < right) {
            if (leftMax < rightMax) {
                left++;
                leftMax = max(leftMax, height[left]);
                water += leftMax - height[left];
            } else {
                right--;
                rightMax = max(rightMax, height[right]);
                water += rightMax - height[right];
            }
        }

        return water;
    }
};

